### PR TITLE
lib: add new lib `event_queue` a data-structure to manage queue of events

### PR DESCRIPTION
**Pull request by privat**
The objective is to be simple to use, powerful, efficient and flexible.

* simple to use:  no framework, only two public concrete classes, a lot of documentation
* powerful: instant, infinite and timed events, dynamic addition and removal of events, a lot of event information (time, duration, completion ration) , fluent-programming-capable
* efficient: internal use of heap and deadlines
* flexible: agnostic of time units, generic (use whatever you want in your events)

It initially targeted the management of the orchestration of graphical effects in a videogame, but become more general and useful.

Signed-off-by: Jean Privat <jean@pryen.org>

### CODE REVIEW

File: lib/event_queue.nit [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	# Ratio of completion
+	#
+	# Usually between 0.0 and 1.
+	# It might be > 1.0 if the event expired during the time-frame.
+	var completion = 0.0
+
+	# Is the event dead?
+	#
+	# Such an event will be not present in the next `update`.
+	var is_dead = false
```
Comment (xymus) [inline lib/event_queue.nit]: I don't like this method name... why not `is_expired`? It is more precise and it's the expression used in all the related documentation.

**Comment by xymus** [comment]
+1

File: lib/event_queue.nit (hunk 1 above)
Comment (privat) [inline lib/event_queue.nit]: If I rename `is_dead` I should also rename `kill`. A proposal? `force_expire`?

Comment (xymus) [inline lib/event_queue.nit]: Yes, force_expire is alright... Force_expiry would work too.

**Comment by privat** [comment]
reroll

* rename `is_dead` as `has_expired` and `kill` as `expire` (note that google fight and @R4PaSs prefer *has expired* to *is expired*)
* add `occurrences` to count how many times an event is returned (it can be used to know when this is the first time we see an event for instance)
* some doc improvements

File: lib/event_queue.nit [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	fun add_at(event: E, start: Float, duration: nullable Float): EventInfo[E]
+	do
+		if start < next then next = start
+		var ei = new EventInfo[E](event, self, start, duration or else 0.0)
+		queue.add ei
+
+		return ei
+	end
+
+	# Add a given amount of time and returns all the event that where active during the delay.
```
Comment (Morriar) [inline lib/event_queue.nit]: "and return~~s~~"

Comment (Morriar) [inline lib/event_queue.nit]: s/where/were

File: lib/event_queue.nit [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+
+	# Register an event that starts after the end of the current one.
+	#
+	# `delay` indicates the time between the end of the current event and the begin of the new one.
+	# Use 0.0 if both events should be contiguous and not overlap.
+	#
+	# Returns the new event information that can be used in fluent programming.
+	fun add_after(event: E, delay: Float, duration: nullable Float): EventInfo[E]
+	do
+		return queue.add_at(event, self.start + self.duration + delay, duration)
```
Comment (Morriar) [inline lib/event_queue.nit]: ~~self.~~start ?

**Comment by Morriar** [comment]
+1

**Comment by privat** [comment]
rereroll: apply suggestions of @Morriar and improve the *fluent programming* part by using a real fluent programming example (as suggested again by Morriar)

**Pull request merged**